safe_autonomy_simulation/dynamics/ode.py

```diff
@@ -93,143 +97,88 @@ def compute_state_dot(
-                    t_eval=t_eval,
-                )
-
-                self.trajectory = sol.y.T
-                self.trajectory_t = sol.t
-
-                next_state = sol.y[:, -1]  # save last timestep of integration solution
-                state_dot = self.compute_state_dot(step_size, next_state, control)
-            else:
-                assert (
-                    self.trajectory_samples <= 0
-                ), "trajectory sampling not currently supported with rk45 jax integration"
-
-                sol = jode.odeint(  # pylint: disable=used-before-assignment
-                    self.compute_state_dot_jax,
-                    state,
-                    np.linspace(0.0, step_size, 11),
-                    control,
-                )
-                next_state = sol[-1, :]  # save last timestep of integration solution
-                state_dot = self.compute_state_dot(step_size, next_state, control)
-        elif self.integration_method == "Euler":
-            assert (
-                self.trajectory_samples <= 0
-            ), "trajectory sampling not currently supported with euler integration"
-            state_dot = self.compute_state_dot(0, state, control)
-            next_state = state + step_size * state_dot
-        else:
-            raise ValueError(f"invalid integration method '{self.integration_method}'")
-
-        return next_state, state_dot
-
-    def compute_state_dot_jax(self, state: np.ndarray, t: float, control: np.ndarray):
-        """Compute state dot for jax odeint
+    def _compute_state_dot(
+        self,
+        t: float,
+        state: np.ndarray | jnp.ndarray,
+        control: np.ndarray | jnp.ndarray,
+    ) -> np.ndarray | jnp.ndarray:
+        raise NotImplementedError
 
-        Jax requires a specific signature for the function passed to odeint. This function
-        is a wrapper around the compute_state_dot function that has the correct signature.
+    def _step(
+        self,
+        step_size: float,
+        state: np.ndarray,
+        control: np.ndarray,
+    ) -> typing.Tuple[np.ndarray, np.ndarray]:
+        if not self.use_jax:  # use numpy arrays and functions
+            step_fn = (
+                safe_autonomy_simulation.dynamics.utils.step_rk45
+                if self.integration_method == "RK45"
+                else safe_autonomy_simulation.dynamics.utils.step_euler
+            )
+        else:  # use jax arrays and JIT functions
+            step_fn = (
+                safe_autonomy_simulation.jax.ode.step_rk45
+                if self.integration_method == "RK45"
+                else safe_autonomy_simulation.jax.ode.step_euler
+            )
+            state = jnp.array(state)
+            control = jnp.array(control)
+
+        next_state, state_dot, self.trajectory, self.trajectory_t = step_fn(
+            f=self.compute_state_dot,
+            step_size=step_size,
+            state=state,
+            control=control,
+        )
```

Comment:
In the case where `use_jax` is true, we could convert the jnp array back to a np array after the JIT function call so that jnp isn't being used outside of the JIT function.  My worry would be that we start with a numpy array, convert to jax numpy, and then the state just stays as a jax numpy array the rest of the time.  It seems like jax numpy is slower than numpy if you're not using JIT or offloading to some accelerator.  If you find that converting back to numpy doesn't really overcome the initial conversion to jax numpy, I'm good with leaving it out.

Reply:
I think it's something we should investigate in a separate issue. It's a good point.

Reply:
Sounds good.  I'll open a new issue.  Marking this resolved.